### PR TITLE
Upgrade packages and fix local development environment

### PR DESCRIPTION
**Pull request by InoMurko**
:clipboard: Add associated issues, tickets, docs URL here.

## Overview

Describe what your Pull Request is about in a few sentences.

1. Updated hardhat package version to `2.9.3` for all packages
2. Remove the builder image by introducing `Dockerfile.packages` to place it 
3. Use the local compiler in `Dockerfile.packages`
4. Remove `run.js` file and use `ts-node run.ts` to start the service
5. Fix the compiling issues for Mac with the M1 chip
6. Fix Bobalink contracts
7. Run Bobalink service in docker for integration tests
8. Update the building process in the release process

## Changes

Describe your changes and implementation choices. More details make PRs easier to review.

- To standardize the hardhat version that we use for different packages, we pick the latest version and upgrade all services to it
- The builder image is removed in `ops/docker-compose.yml` to speed up. Almost all services can be built from the `Dockerfile.packages` file
- The hardhat's latest version can block the Mac with the M1 chip from downloading the Solidity compiler. To solve the problem, we download the compiler first and move it to an appropriate folder so that hardhat can use the local compiler.
- To simplify our codebase, the code for starting the service is changed to `ts-node run.ts`
- M1 Mac can't compile solidity contracts in docker due to the system version. It's solved in `Dockerfile.packages` file.
- Fix the problem that only owner can call the HC function. The modifier should be `oracleOwner`.
```
  function getChainLinkQuote(uint256 _roundId)
    public
    onlyOracleAdmin
    returns (uint256 , int256, uint256)
  {}
```
- Fix the problem of the wrong op code in the Bobalink contract.
```
    address[] memory oracleAddresses = new address[](1);
    oracleAddresses[0] = oracleAddress;
    return oracleAddresses;
```
- Add the initialization function for Bobalink contracts in `packages/boba/contracts` in order to start Bobalink service in docker successfully
- Remove `build-ci.sh` and use `Dockerfile.packages` to build images for the release

## Testing

Describe how to test your new feature/bug fix and if possible, a step by step guide on how to demo this.

The unit tests are added for new features

### CODE REVIEW

File: packages/boba/bundler/yarn.lock [hunk 1]
```diff
@@ -140,6 +140,37 @@
     minimatch "^3.1.2"
     strip-json-comments "^3.1.1"
 
+"@eth-optimism/contracts@^0.5.11":
```
Comment (boyuan-chen) [inline packages/boba/bundler/yarn.lock]: This yarn.lock can be removed?

File: packages/boba/contracts/package.json [hunk 2]
```diff
@@ -18,42 +18,46 @@
     "test": "yarn test:contracts",
     "test:contracts": "hardhat test --show-stack-traces",
     "test:coverage": "NODE_OPTIONS=--max_old_space_size=8192 hardhat coverage",
-    "test:slither": "slither ."
+    "test:slither": "slither .",
+    "depcheck": "depcheck --ignores @openzeppelin/*,@chainlink/*"
```
Comment (InoMurko) [inline packages/boba/contracts/package.json]: there's a .deepcheck file!

Comment (boyuan-chen) [reply packages/boba/contracts/package.json]: Added .depcheckrc file for boba/contracts

File: integration-tests/test/bobalink.spec.ts [hunk 3]
```diff
@@ -331,90 +295,164 @@ describe('BobaLink Test\n', async () => {
...
+
+  it('should get a single quote via bobalink using test api', async () => {
+    await EthOracleHC.updateHCUrl(`${URL}/bobalink-test-api`)
+    const lastRoundId = (await EthChainLinkOracle.latestRound()).toNumber()
+    const decimals = await EthChainLinkOracle.decimals()
+    const price = utils.parseUnits('12000', decimals)
+    /* eslint-disable */
+    await EthChainLinkOracle.emergencySubmit(lastRoundId + 1, price, 1000)
+    /* eslint-enable */
+    await sleep(10000)
```
Comment (InoMurko) [inline integration-tests/test/bobalink.spec.ts]: sleep is a bad pattern to be used. we should always use pull!
pull every N seconds for X amount of time

File: ops/docker-compose-side.yml [hunk 4]
```diff
@@ -66,6 +74,13 @@ services:
...
-      dockerfile: ./ops/docker/Dockerfile.bobalink
-    deploy:
-      replicas: 0
+      dockerfile: ./ops/docker/Dockerfile.packages
+      target: bobalink
+    environment:
+      L1_NODE_WEB3_URL: http://l2geth:8545
+      L2_NODE_WEB3_URL: http://l2geth:8545
+      ADDRESS_MANAGER_ADDRESS: "0x5FbDB2315678afecb367f032d93F642f64180aa3"
+      << : *bobalink_pk #same as the deployer
```
Comment (InoMurko) [inline ops/docker-compose-side.yml]: is this wise? what if we use a dedicated PK?

**Review by wsdt — APPROVED** [review]

File: ops/docker-compose-side.yml (hunk 4 above)
Comment (souradeep-das) [reply ops/docker-compose-side.yml]: yeah account 14 is probably also used on tests? also can we fix this comment?

File: packages/boba/contracts/deploy/022-PriceFeedOracleHC.deploy.ts [hunk 5]
```diff
@@ -93,6 +91,16 @@ const deployFn: DeployFunction = async (hre) => {
         FluxAggregatorHCDeploymentSubmission
       )
 
+      await FluxAggregatorHC.initialize(
```
Comment (souradeep-das) [inline packages/boba/contracts/deploy/022-PriceFeedOracleHC.deploy.ts:94]: can we list this under PR changes? and the base-service changes too?

File: .github/workflows/boba-contracts-unit.yml [hunk 6]
```diff
@@ -1,39 +0,0 @@
...
-      - 'master'
-      - 'develop'
-      - '*rc'
-      - 'regenesis/*'
-  pull_request:
-  workflow_dispatch:
-
-jobs:
-
-  boba-unit-tests:
```
Comment (souradeep-das) [inline .github/workflows/boba-contracts-unit.yml:15]: we wont run this?

**Review by souradeep-das — APPROVED** [review]

File: packages/boba/contracts/deploy/022-PriceFeedOracleHC.deploy.ts (hunk 5 above)
Comment (boyuan-chen) [reply packages/boba/contracts/deploy/022-PriceFeedOracleHC.deploy.ts:94]: Sure! I will write up all changes once everything is settle down.

File: .github/workflows/boba-contracts-unit.yml (hunk 6 above)
Comment (boyuan-chen) [reply .github/workflows/boba-contracts-unit.yml:15]: The tests are run in the coverage tests.

File: ops/docker-compose-side.yml (hunk 4 above)
Comment (boyuan-chen) [reply ops/docker-compose-side.yml]: Replaced to a new key!

File: integration-tests/test/bobalink.spec.ts (hunk 3 above)
Comment (boyuan-chen) [reply integration-tests/test/bobalink.spec.ts]: Added this
```
export const waitForAndExecute = async (
  fn: () => Promise<any>,
  repeat: number,
  intervalMs?: number,
) => {
  while (repeat >= 0) {
    try {
      await fn()
      break
    } catch (e) {
      repeat = repeat - 1
      await sleep(intervalMs || 1000)
    }
  }
}
```

File: packages/boba/contracts/deploy/022-PriceFeedOracleHC.deploy.ts (hunk 5 above)
Comment (boyuan-chen) [reply packages/boba/contracts/deploy/022-PriceFeedOracleHC.deploy.ts:94]: Added!

**Comment by InoMurko** [comment]
Thank you Boyuan for updating the description on the PR. I think we're good to go.

File: packages/boba/contracts/contracts/oracle/FluxAggregatorHC.sol [hunk 7]
```diff
@@ -220,7 +220,7 @@ contract FluxAggregatorHC is AggregatorV2V3Interface {
    * @notice returns the oracle address
    */
   function getOracles() external view returns (address[] memory) {
-    address[] memory oracleAddresses;
+    address[] memory oracleAddresses = new address[](1);
```
Comment (InoMurko) [inline packages/boba/contracts/contracts/oracle/FluxAggregatorHC.sol:223]: boom! amazing cc @souradeep-das this didn't work

Comment (souradeep-das) [reply packages/boba/contracts/contracts/oracle/FluxAggregatorHC.sol:223]: ohh makes sense

**Pull request merged**